File: Ray Tracing in One Weekend/util/vec3.py
import math
import random
import numpy as np

class Vec3:
    def __init__(self, x=0, y=0, z=0):
        self.e = [x, y, z]

    # --------------------------
    # Properties
    # --------------------------
    @property
    def x(self):
        return self.e[0]

    @property
    def y(self):
        return self.e[1]

    @property
    def z(self):
        return self.e[2]

    # --------------------------
    # Operators
    # --------------------------
    def __neg__(self):
        return Vec3(-self.x, -self.y, -self.z)

    def __add__(self, other):
        return Vec3(
            self.x + other.x,
            self.y + other.y,
            self.z + other.z
        )

    def __sub__(self, other):
        return Vec3(
            self.x - other.x,
            self.y - other.y,
            self.z - other.z
        )

    def __mul__(self, t):
        if isinstance(t, Vec3):
            return Vec3(
                self.x * t.x,
                self.y * t.y,
                self.z * t.z
            )

        return Vec3(
            self.x * t,
            self.y * t,
            self.z * t
        )

    def __rmul__(self, t):
        return self.__mul__(t)

    def __truediv__(self, t):
     # CHANGE: avoid divide-by-zero crashes in normalization
     return self * (1 / (t if t != 0 else 1e-12))

    def __str__(self):
        return f"{self.x} {self.y} {self.z}"

    # --------------------------
    # Length
    # --------------------------
    def length_squared(self):
        return (
            self.x * self.x +
            self.y * self.y +
            self.z * self.z
        )

    def length(self):
        return math.sqrt(self.length_squared())

    # --------------------------
    # NEW: Near-zero check
    #
    # Purpose:
    # Detect vectors very close to (0,0,0)
    #
    # Formula:
    #
    # |x| < ε
    # |y| < ε
    # |z| < ε
    #
    # where ε = 10^-8
    # --------------------------
    def near_zero(self):

        s = 1e-8

        return (
            abs(self.x) < s
            and abs(self.y) < s
            and abs(self.z) < s
        )

    # --------------------------
    # Random vectors
    # --------------------------
    @staticmethod
    def random():
        return Vec3(
            random.random(),
            random.random(),
            random.random()
        )

    @staticmethod
    def random_range(min_v, max_v):
        return Vec3(
            random.uniform(min_v, max_v),
            random.uniform(min_v, max_v),
            random.uniform(min_v, max_v)
        )


# --------------------------
# Utility functions
# --------------------------
def dot(u, v):
    return (
        u.x * v.x +
        u.y * v.y +
        u.z * v.z
    )


def cross(u, v):
    return Vec3(
        u.y * v.z - u.z * v.y,
        u.z * v.x - u.x * v.z,
        u.x * v.y - u.y * v.x
    )


#def unit_vector(v):
#    return v / v.length()


# --------------------------
# Diffuse helpers
# --------------------------
def random_unit_vector():
    """
    Rejection sampling inside unit sphere
    """

    while True:
        p = Vec3.random_range(-1, 1)

        lensq = p.length_squared()

        if 1e-160 < lensq <= 1:
            return p / math.sqrt(lensq)


def random_on_hemisphere(normal):
    """
    Keep ray in outward hemisphere
    """

    on_unit_sphere = random_unit_vector()

    if dot(on_unit_sphere, normal) > 0.0:
        return on_unit_sphere

    return -on_unit_sphere

# =========================
# Unit vector (normalization)
# =========================
def unit_vector(v):
    len_v = v.length()

    # CHANGE:
    # prevents NaN when vector is zero
    if len_v < 1e-12:
        return Vec3(0, 0, 0)

    return v / len_v


# =========================
# REFLECTION FUNCTION
# =========================
def reflect(v, n):
    """
    PURPOSE:
    Mirror reflection of vector v around normal n

    FORMULA:
        r = v - 2(v·n)n

    WHY:
    - removes normal component twice
    - flips direction symmetrically
    """

    return v - 2 * dot(v, n) * n

def refract(uv, n, etai_over_etat):
    """
    uv: unit direction of incoming ray
    n : surface normal
    etai_over_etat: ratio of refractive indices (η / η')
    """

    # STEP 1: compute cos(theta)
    # CHANGED: we use dot product instead of angle
    cos_theta = min((-uv).dot(n), 1.0)

    # STEP 2: perpendicular component (bending part)
    # CHANGED: scales direction using Snell’s Law form
    r_out_perp = (uv + n * cos_theta) * etai_over_etat

    # STEP 3: parallel component (keeps unit length)
    # CHANGED: ensures physical constraint |R'| = 1
    r_out_parallel = n * (-math.sqrt(abs(1.0 - r_out_perp.length_squared())))

    # STEP 4: final refracted ray
    return r_out_perp + r_out_parallel

Point3 = Vec3
Color = Vec3

File: Ray Tracing in One Weekend - Javascript/Ch12-B/vec3.js
class Vec3{
    constructor(x,y,z){
        this.e=[x,y,z]
        
    }


    get x() { return this.e[0]; }
    set x(v) { this.e[0] = v; }

    get y() { return this.e[1]; }
    set y(v) { this.e[1] = v; }

    get z() { return this.e[2]; }
    set z(v) { this.e[2] = v; }
    
    neg(){ return new Vec3(-this.e[0],-this.e[1],-this.e[2])}
    get(i){ return this.e[i]}

    add_eq(v){
        this.e[0] += v.e[0];
        this.e[1] += v.e[1];
        this.e[2] += v.e[2]; 
        return this       
    }

    mul_eq(t){
        this.e[0] *= t
        this.e[1] *= t
        this.e[2] *= t   
        return this     
    }

    div_eq(t){
        this.mul_eq(1.0/t)
        return this
    }

    length_squared(){
        return this.e[0]*this.e[0] + this.e[1]*this.e[1] + this.e[2]*this.e[2];
    }

    near_zero()  {
        // Return true if the vector is close to zero in all dimensions.
        var s = 1e-8;
        return (Math.abs(this.e[0]) < s) && (Math.abs(this.e[1]) < s) && (Math.abs(this.e[2]) < s);
    }

    length(){
        return Math.sqrt(this.length_squared())
    }


    static add(u,v){
        if (!u || !v) {
         return new Vec3(0,0,0);
    }
        return new Vec3(u.e[0] + v.e[0], u.e[1] + v.e[1], u.e[2] + v.e[2]);
    }

    static sub(u,v){
    
        return new Vec3(u.e[0] - v.e[0], u.e[1] - v.e[1], u.e[2] - v.e[2]);
    }

    static mul({t=1,u = new Vec3(1,1,1),v}){
        return new Vec3(t*u.e[0] * v.e[0], t*u.y * v.e[1], t*u.z * v.e[2]);
    }
 
    static div(v,t){
        return this.mul({t:1.0/t,v:v})
    }

    static dot(u,v){
        return u.e[0] * v.e[0]
         + u.e[1] * v.e[1]
         + u.e[2] * v.e[2]
    }
    
    static cross(u,v){
        return new Vec3(u.e[1] * v.e[2] - u.e[2] * v.e[1],
                u.e[2] * v.e[0] - u.e[0] * v.e[2],
                u.e[0] * v.e[1] - u.e[1] * v.e[0])
    }

    static unit_vector(v){
        return this.div(v,v.length())
    }

    static random_double(min, max) {
    return min + (max - min) * Math.random();
    }

    static random(){
        return new Vec3(Math.random(),Math.random(),Math.random())
    }

    static random(min, max){
        return new Vec3(this.random_double(min,max),this.random_double(min,max),this.random_double(min,max))
    }


    static random_unit_vector(){
        while(true){
            var p = this.random(-1.0,1.0)
            var lensq = p.length_squared()
            if((1e-160 < lensq) && (lensq <= 1 ))
                return Vec3.div(p, Math.sqrt(lensq));
        }
    }

    static random_on_hemisphere(N){
        var on_unit_sphere = this.random_unit_vector()
        return (this.dot(on_unit_sphere,N)>0.0) ? on_unit_sphere : on_unit_sphere.neg()
    }

    static reflect( v, n) {
        //    return v - 2*dot(v,n)*n;
        var a = this.mul({t:2*Vec3.dot(v,n),v:n})
        return Vec3.sub(v , a);
    }

    static refract(uv, n,etai_over_etat ){
        var cos_theta = Math.min(this.dot(uv.neg(),n),1.0)
        var r_out_perp = this.mul({t:etai_over_etat  ,v: this.add(uv , this.mul({t:cos_theta,v:n}))})
        var r_out_parallel = this.mul({t:-Math.sqrt(Math.abs(1.0-r_out_perp.length_squared())) , v:n})
        return this.add(r_out_perp,r_out_parallel)
    }

    static degrees_to_radians = degrees => degrees * (Math.PI / 180);
}
// point3 is just an alias for vec3, but useful for geometric clarity in the code.
window.Point3 = Vec3

// Vector Utility Functions
//https://stackoverflow.com/questions/43642729/calling-a-method-from-another-method-in-the-same-class



window.Vec3 = Vec3



